File: src/editor/database/symbol/Symbol.java
package editor.database.symbol;

import java.awt.Image;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.regex.Pattern;

import javax.swing.Icon;
import javax.swing.ImageIcon;

/**
 * This class represents a symbol that might appear on a card in Magic: The Gathering.  It has a weight for each
 * color, which is equal to 1 divided by the number of colors the symbol has, or zero if it does not have that
 * color.  Symbol cannot be instantiated; rather, instances of it are accessed using static variables that contain
 * all possible values those symbols can attain.  For colored symbols, those variables will be Maps mapping colors
 * onto the correct symbols.  For constant generic symbols, they will be an array containing the colorless symbol
 * corresponding to the index into the array and others for special symbols (100 and 1,000,000).  Nonmana symbols,
 * variable symbols, and the half-mana symbol all have a single static variable that contains an instance of that
 * class.
 * 
 * @author Alec Roelke
 */
public abstract class Symbol
{
	/**
	 * Pattern for finding an individual symbol in a string.
	 */
	public static final Pattern SYMBOL_PATTERN = Pattern.compile("\\{([^}]+)\\}");
	/**
	 * Name for a Symbol whose icon file can't be found.
	 */
	public static final String UNKNOWN_ICON = "unknown.png";
	
	/**
	 * Create a Symbol from a String.
	 * 
	 * @param s String representation of the new symbol, not surrounded by {}
	 * @return a new symbol that the specified String represents
	 * @throws IllegalArgumentException if the String doesn't correspond to a symbol
	 */
	public static Symbol parseSymbol(String s) throws IllegalArgumentException
	{
		Symbol symbol = tryParseSymbol(s);
		if (symbol == null)
			throw new IllegalArgumentException('"' + s + "\" is not a symbol");
		return symbol;
	}
	
	/**
	 * Create a Symbol from a String.
	 * 
	 * @param s String representation of the new symbol, not surrounded by {}
	 * @return a new symbol that the specified String represents, or null if there is
	 * no such symbol.
	 */
	public static Symbol tryParseSymbol(String s)
	{
		Symbol value;
		if ((value = ManaSymbol.tryParseManaSymbol(s)) != null)
			return value;
		else if ((value = FunctionalSymbol.tryParseFunctionalSymbol(s)) != null)
			return value;
		else
			return null;
	}
	
	/**
	 * Icon to show when displaying this Symbol.
	 */
	private final ImageIcon icon;
	/**
	 * Name of the file containing the icon (not including parent directory).
	 */
	private final String name;
	/**
	 * The textual representation of this Symbol.
	 */
	private final String text;
	
	/**
	 * Create a new Symbol.
	 * 
	 * @param iconName name of the icon file for the new Symbol
	 */
	protected Symbol(String iconName, String t)
	{
		if (Files.notExists(Paths.get("images/icons/" + iconName)))
		{
			System.err.println("Could not load file images/icons/" + iconName);
			iconName = UNKNOWN_ICON;
		}
		icon = new ImageIcon("images/icons/" + (name = iconName));
		text = t;
	}
	
	@Override
	public boolean equals(Object other)
	{
		return other != null && (other == this || other.getClass() == getClass() && toString().equals(other.toString()));
	}
	
	/**
	 * Get the icon that should be used to display this Symbol.
	 * 
	 * @return this Symbol's icon with its default width.
	 */
	public Icon getIcon()
	{
		return icon;
	}
	
	/**
	 * Get this Symbol's icon with the specified width.  The height will be scaled accordingly as well.
	 * 
	 * @param newSize width of the icon
	 * @return the resized icon
	 */
	public Icon getIcon(int newSize)
	{
		return new ImageIcon(icon.getImage().getScaledInstance(-1, newSize, Image.SCALE_SMOOTH));
	}
	
	/**
	 * Get the name of this Symbol.
	 * 
	 * @return the name of this Symbol
	 */
	public String getName()
	{
		return name;
	}

	@Override
	public int hashCode()
	{
		return toString().hashCode();
	}

	/**
	 * {@inheritDoc}
	 * The String representation of this Symbol is its text surrounded by {}.
	 */
	@Override
	public String toString()
	{
		return "{" + text + "}";
	}
}

File: src/main/scala/editor/database/symbol/Symbol.scala
package editor.database.symbol

import java.awt.Image
import java.io.IOException
import javax.imageio.ImageIO
import javax.swing.ImageIcon
import scala.util.matching._

/**
 * An object that is capable of parsing a [[Symbol]] from a string.
 * 
 * @tparam S type of [[Symbol]] being parsed
 * 
 * @author Alec Roelke
 */
trait SymbolParser[S <: Symbol] {
  /**
   * Parse a [[Symbol]] from a string.
   * 
   * @param s string to parse
   * @return the [[Symbol]] corresponding to the string
   */
  def parse(s: String): Option[S]
}

/**
 * An object that contains a set of distinct symbols, indexed by some key.
 * 
 * @tparam K key used for indexing
 * @tparam S type of symbol contained
 * 
 * @author Alec Roelke
 */
trait HasSymbolValues[K, S <: Symbol] {
  /** Mapping of keys onto their respective [[Symbol]]s. */
  def values: Map[K, S]

  /**
   * Shortcut for [[values.apply]] to get the symbol for a particular key.
   * 
   * @param key key of the [[Symbol]] to get
   * @return the [[Symbol]] corresponding to the key
   */
  def apply(key: K) = values(key)
}

/**
 * Companion to [[Symbol]] containing global data and methods for parsing from strings.
 * @author Alec Roelke
 */
object Symbol extends SymbolParser[Symbol] {
  /** Regular expression for retrieving a [[Symbol]]'s text from its string representation. */
  val Regex = raw"\{([^}]+)\}".r

  /** Icon name to use for [[Symbol]]s whose icons are unknown. */
  val Unknown = "unknown.png"

  /**
   * Parse a [[Symbol]] from a string. The string should be its text representation without braces.
   * 
   * @param s string to parse
   * @return the [[Symbol]] represented by the string, or None if there isn't one
   */
  override def parse(s: String) = ManaSymbol.parse(s) orElse FunctionalSymbol.parse(s)
}

/**
 * A symbol that can appear on a card. Since there is a finite set of symbols, individual ones cannot be instantiated;
 * rather, they are accessed via static data structure or by parsing from string.
 * 
 * @constructor create a new symbol
 * @param name file name containing the icon to show for the symbol
 * @param text text representation of the symbol; when printed out, this will be displayed between braces
 * 
 * @author Alec Roelke
 */
abstract class Symbol(val name: String, private val text: String) {
  val icon = try {
    ImageIcon(ImageIO.read(getClass.getResourceAsStream(s"/images/icons/$name")))
  } catch case e: IOException => { e.printStackTrace(); ImageIcon() }

  def scaled(size: Int) = ImageIcon(icon.getImage.getScaledInstance(-1, size, Image.SCALE_SMOOTH))

  override def toString = s"{$text}"
  override def hashCode = toString.hashCode
  override def equals(other: Any) = other match {
    case null => false
    case ref: AnyRef if ref eq this => true
    case _ => other.getClass == getClass && other.toString == toString
  }
}